rir/src/compiler/translations/rir_2_pir_compiler.h

```diff
@@ -0,0 +1,47 @@
+#ifndef RIR_2_PIR_COMPILER_H
+#define RIR_2_PIR_COMPILER_H
+
+#include "../opt/cleanup.h"
+#include "../opt/delay_env.h"
+#include "../opt/delay_instr.h"
+#include "../opt/elide_env.h"
+#include "../opt/force_dominance.h"
+#include "../opt/inline.h"
+#include "../opt/scope_resolution.h"
+#include "rir_compiler.h"
+
+namespace rir {
+namespace pir {
+
+class Rir2PirCompiler : public RirCompiler {
+  public:
+    Rir2PirCompiler(Module* module) : RirCompiler(module) {
+        translations.push_back(new ForceDominance());
```

Comment:
personally I would have used lambdas here. Let me show you an alternative implementation:

```
class Rir2PirCompiler {
  typedef std::function<Closure*(Closure*)> PirTranslatorFunction;
  std::vector<PirTranslatorFunction> translations;
  ...
    // register a translation:
    translations.push_back([](Closure* c){return ForceDominance::apply(c);});
  ...
    // apply translation:
    Closure* cur = ...;
    for (translate : translations) {
        cur = translate(cur);
    }
```

This approach give you much less coupling! You can get away without all the translators implementing the exact same interface. You don't have dependencies in the header files. Your don't need virtual methods. And you don't need to instantiate the translators, like here, but can instead use static methods, like before.